File: abstractions/typescript/src/serialization/parsable.ts
import { ParseNode } from "./parseNode";
import { SerializationWriter } from "./serializationWriter";

/**
 * Defines a serializable model object.
 */
export interface Parsable {
  /**
   * Gets the deserialization information for this object.
   * @return The deserialization information for this object where each entry is a property key with its deserialization callback.
   */
  getFieldDeserializers<T>(): Record<string, (item: T, node: ParseNode) => void>;
  /**
   * Writes the objects properties to the current writer.
   * @param writer The writer to write to.
   */
  serialize(writer: SerializationWriter): void;
  /**
   * Gets the additional data for this object that did not belong to the properties.
   * @return The additional data for this object.
   */
  additionalData: Record<string, unknown>;
}

File: abstractions/go/serialization/parsable.go
package serialization

// Parsable defines a serializable model object.
type Parsable interface {
	// Serialize writes the objects properties to the current writer.
	Serialize(writer SerializationWriter) error
	// GetFieldDeserializers returns the deserialization information for this object.
	GetFieldDeserializers() map[string]func(interface{}, ParseNode) error
}

// ParsableFactory is a factory for creating Parsable.
type ParsableFactory func(parseNode ParseNode) (Parsable, error)
